proposal/q40603/proposal.md

```diff
@@ -0,0 +1,153 @@
+![](https://i.imgur.com/q0DDp8O.png)
+
+
+Select pairs by cointegration approach:
+1. unit-root test
+    ADF test
+2. information criteria 
+    BIC test
+5. Johansen test for cointegration
+6. Likelihood ratio test for selecting adequate VECM model
+
+
+
+
+
+#### What I want to imporve
+
+* The structure of the code
+* CI procedure
+* The computaion speed
+
+
+
+### Perspective users
+
+
+Trader or the students under department of Finance.
+
+### System architecture
+
+input be like:
+
+stock table
+| | 2330 (TSMC) | 2454(MTK) |... |
+| -------- | -------- | -------- | -------- |
+| 1st min average price     | 612     | 933     |...     |
+| 2nd min average price     | 614     | 930     |...     |
+
+1. Stock raw tick data preprocessing
+2. VAR model test
+3. portmanteau test
+4. Normality test
+5. model selection
+6. calculate weight of two stocks
+7. output
+
+pairs table
+| id | stock1 | stock2 | weight 1 | weight2 | mean | standard deviation |
+| -------- | -------- | -------- | -------- |-------- |-------- |-------- |
+| 1| 2330     | 2454     | 0.3...    | 0.6... | 2.3... |  0.1... |
+|  2| 2317     | 2308     | 0.1...     | 0.8... |  1.7... | 0.2... |
+
+
+### API description
+
+
+Python : top-level API wrapped from C++ code
+
+
+| Name | input | return |description |
+| -------- | -------- | -------- |  -------- |
+| cal_pairs     |  stock table (pandas object)  | pairs table  |  user may import the function input the table mentioned above to do the calculation  |   |
+| back_test | pairs table| profit, win rate  | user may calculate the profit with the function  | 
+
+
+C++ : 
+| Name | input | return |description |
+| -------- | -------- | -------- |  -------- |
+| ADF_test     |  series  | bool  |  check if the series is stationary  |   |
+| VAR | series  | return model object |vector autoregrssive model, there are some existing library for this like statsmodels.tsa.api |
+| order_select | series  |  int  | select best BIC order  |
+| rank | data,model,p | int | test for each model |
+
+There should be more, but I am not sure what is the best way to modularize it.
+
+
+### Engineering infrastructure
+
+The basic goal of this project is to speed up the original python version work.
+Since there are many vector manipulation, I will make use of MKL library and also other existing C++ statisic libraries. The final goal is to make the user import the function from python to use. I will also add some Unit test and even CI to this project. Since this is a huge project (in my opinion), I will make use of TDD principles and do whatever the best I can do.
+
+### Schedule
+
+| date | work | comment |
+| -------- | -------- | -------- |
+| 4/7 ~ 4/30   | Input data processing, modularize the computation (design part)    |
+| 5/1 ~ 5/15 | ADF test ,VAR model, order select |
+| 5/15 ~ presentation| VECM weight calcualtion | 
+
+
+
+### Question
+
+1. In Numerical Software Development, are there any special design pattern for us to follow?
```

Comment:
This is an excellent question.  It's no and yes.  No, the design patterns for numerical, scientific, and engineering codes are are special.  But yes, the codes do use a special selection and combination of the patterns.

There is not a good reference for the numerical design patterns, but I think the patterns used in game design have good resemblance.  You may take a look at https://gameprogrammingpatterns.com/contents.html .